sklearn/utils/estimator_checks.py

```diff
@@ -3301,15 +3302,16 @@ def _enforce_estimator_tags_x(estimator, X):
     # Pairwise estimators only accept
     # X of shape (`n_samples`, `n_samples`)
     if _safe_tags(estimator, key="pairwise"):
-        X = X.dot(X.T)
+        if X.shape[0] != X.shape[1] or not np.allclose(X, X.T):
+            X = X.dot(X.T)
```

Comment:
I see that this is a suggestion from https://github.com/scikit-learn/scikit-learn/pull/24136/files#r974671132. I think the better solution is to remove `_pairwise_estimator_convert_X` itself and have the logic go into `_enforce_estimator_tags_x`. To keep this PR small, I suggest not changing this PR.

To remind ourselves to fix this, can we leave a comment here?

```suggestion
        # TODO: Remove when `_pairwise_estimator_convert_X` is removed and
        # the it's functionality is moved here
        if X.shape[0] != X.shape[1] or not np.allclose(X, X.T):
            X = X.dot(X.T)
```